### PR TITLE
Allow event creation

### PR DESCRIPTION
**Pull request by krichprollsch**

### CODE REVIEW

File: src/Api/EventApi.php [hunk 1]
```diff
@@ -158,5 +159,27 @@ public function get($identifier)
...
+
+        $method = 'POST';
+        $url = sprintf('/me/calendars/%s/events', $event->getCalendar()->getId());
+        if (null !== $event->getId()) {
+            $method = 'PATCH';
+            $url = sprintf('%s/%s', $url, $event->getId());
+        }
+
+        $request = $this->guzzle->createRequest($method, $url, $params);
+        $response = $this->guzzle->send($request);
```
Comment (Taluu) [inline src/Api/EventApi.php:179]: why don't you use `$this->guzzle->post()` directly ?

**Comment by Taluu** [comment]
lgtm :+1:

File: src/Api/EventApi.php (hunk 1 above)
Comment (krichprollsch) [inline src/Api/EventApi.php:179]: I just base my guzzle usage with the other methods

**Comment by krichprollsch** [comment]
@Taluu I did some update, could-you re-check plz ?

File: src/Model/Event.php [hunk 2]
```diff
@@ -334,4 +334,25 @@ public static function hydrate(array $data, Calendar $calendar = null)
...
         return $event;
     }
+
+    public function export()
+    {
+        $export = [
+            'subject' => $this->getName(),
+            'bodyPreview' => $this->getDescription(),
+            'start' => ['dateTime' => $this->getStart()->format('c'), 'timeZone' => 'UTC'],
+            'end' => ['dateTime' => $this->getEnd()->format('c'), 'timeZone' => 'UTC'],
```
Comment (Taluu) [inline src/Model/Event.php]: careful here, on partial events `getStart` and `getEnd` doesn't _always_ return a date....

Comment (Taluu) [inline src/Model/Event.php]: cf https://github.com/Calendart/GoogleAdapter/commit/a6be41018531cf1804f6c9a101d8fde6bb721d4f

File: src/Model/PartialEvent.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            $export[$property] = $value;
+        }
+        return $export;
+    }
+
+    public static function hydrate(array $data, Calendar $calendar = null)
+    {
+        $event = new static($calendar);
+
+dump($data);
```
Comment (Taluu) [inline src/Model/PartialEvent.php]: _kof kof_

File: src/Model/PartialEvent.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+        isset($data['id']) && ($event->id = $data['id']) && ($event->changedProperties['id'] = true);
+        isset($data['changeKey']) && ($event->etag = $data['changeKey']) && ($event->changedProperties['changeKey'] = true);
+        $event->raw = $data;
+        $event->participations = new ArrayCollection;
+
+dump($event);
```
Comment (Taluu) [inline src/Model/PartialEvent.php]: _kof kof_

File: src/Model/PartialEvent.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    public static function hydrate(array $data, Calendar $calendar = null)
+    {
+        $event = new static($calendar);
+
+dump($data);
+
+        isset($data['id']) && ($event->id = $data['id']) && ($event->changedProperties['id'] = true);
+        isset($data['changeKey']) && ($event->etag = $data['changeKey']) && ($event->changedProperties['changeKey'] = true);
```
Comment (Taluu) [inline src/Model/PartialEvent.php]: heum... a `if (isset(...)) { ... }` would be WAY readable...

**Comment by Taluu** [comment]
To be honest I am not seeing the point of hydrating a `PartialEvent`

File: src/Model/PartialEvent.php (hunk 4 above)
Comment (krichprollsch) [inline src/Model/PartialEvent.php]: oops

File: src/Model/PartialEvent.php (hunk 5 above)
Comment (krichprollsch) [inline src/Model/PartialEvent.php]: yeah, agree

**Comment by krichprollsch** [comment]
`PartialEvent` is easier to manage, I don't want to expose untouched attributes.

**Comment by Taluu** [comment]
On event creating you should expose everything you want that is not null or something like that (or expose everything that should be exposed actually). Not seeing the point of not exposing null when it shouldn't it exposed.

The partial event shouldn't actually be built on a hydrate (this is the api classes mecanism)... You should build it on a classical way (with construct and all)

File: src/Model/PartialEvent.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        parent::__construct($calendar);
+
+        if (null !== $id) {
+            $this->id = $id;
+            $event->changedProperties['id'] = true;
+        }
+
+        if (null != $etag) {
+            $this->etag = $etag;
+            $event->changedProperties['changeKey'] = true;
```
Comment (Taluu) [inline src/Model/PartialEvent.php]: should be `$this`

File: src/Model/PartialEvent.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    /** @var array */
+    private $changedProperties = [];
+
+    public function __construct(Calendar $calendar, Datetime $start, Datetime $end, $id = null, $etag = null)
+    {
+        parent::__construct($calendar);
+
+        if (null !== $id) {
+            $this->id = $id;
+            $event->changedProperties['id'] = true;
```
Comment (Taluu) [inline src/Model/PartialEvent.php]: should be `$this`

File: src/Model/PartialEvent.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+    /** {@inheritDoc} */
+    public function addParticipation(BaseEventParticipation $participation)
+    {
+        $this->changedProperties['attendees'] = true;
+        parent::addParticipation($participation);
+    }
+
+    /** {@inheritDoc} */
+    public function export()
```
Comment (Taluu) [inline src/Model/PartialEvent.php:76]: note for later : maybe using a `array_filter` would have been better, but I already bothered @krichprollsch enough as it is

**Comment by Taluu** [comment]
Thanks @krichprollsch

**Pull request merged**